pkg/remote/google/repository/asset.go

```diff
@@ -0,0 +1,81 @@
+package repository
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	asset "cloud.google.com/go/asset/apiv1"
+	"github.com/cloudskiff/driftctl/pkg/remote/cache"
+	"github.com/cloudskiff/driftctl/pkg/remote/google/config"
+	"google.golang.org/api/iterator"
+	assetpb "google.golang.org/genproto/googleapis/cloud/asset/v1"
+)
+
+const (
+	storageBucketAssetType = "storage.googleapis.com/Bucket"
+)
+
+type AssetRepository interface {
+	SearchAllBuckets() ([]*assetpb.ResourceSearchResult, error)
+}
+
+type assetRepository struct {
+	client *asset.Client
+	config config.GCPTerraformConfig
+	cache  cache.Cache
+	lock   sync.Locker
+}
+
+func NewAssetRepository(client *asset.Client, config config.GCPTerraformConfig, c cache.Cache) *assetRepository {
+	return &assetRepository{
+		client,
+		config,
+		c,
+		&sync.Mutex{},
+	}
+}
+
+func (s assetRepository) searchAllResources(ty string) ([]*assetpb.ResourceSearchResult, error) {
+	req := &assetpb.SearchAllResourcesRequest{
+		Scope: fmt.Sprintf("projects/%s", s.config.Project),
+		AssetTypes: []string{
+			storageBucketAssetType,
+		},
+	}
+	var results []*assetpb.ResourceSearchResult
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
```

Comment:
What's the purpose of this locker ?

Reply:
It is to ensure that we do not call `SearchAllResources` multiples times, since this method can be called concurrently we want to avoid to make two call to AWS API if no results are cached initially.